### PR TITLE
Flink: Migrate remaining classes to JUnit5

### PR DESCRIPTION
**Pull request by tomtongue**
Migrate the following classes in `iceberg-flink`  to JUnit 5 for https://github.com/apache/iceberg/issues/9087

  - [x] `TestIcebergSourceBoundedGenericRecord`
  - [x] `TestIcebergSourceContinuous`
  - [x] `TestIcebergSourceFailover`
  - [x] `TestIcebergSourceFailoverWithWatermarkExtractor`
  - [x] `TestIcebergSourceWithWatermarkExtractor`
  - [x] `TestProjectMetaColumn`
  - [x] `TestSourceUtil`

### CODE REVIEW

**Review by nastra — COMMENTED** [review]
changes LGTM, but there are a few more files left in Flink 1.19. Could you please add them in this PR here to so that we can finish Junit5 migration for Flink?

Also I think we can remove `HadoopTableResource` / `HadoopCatalogResource` now.

**Comment by tomtongue** [comment]
> changes LGTM, but there are a few more files left in Flink 1.19. Could you please add them in this PR here to so that we can finish Junit5 migration for Flink?
> 
> Also I think we can remove `HadoopTableResource` / `HadoopCatalogResource` now.

Thanks for checking this. Sure, I believe there are around 5 files left in Flink. Let me add the remaining classes.

**Comment by tomtongue** [comment]
Will remove hadoop catalog classes and add the following tests into this PR:

- [] `SimpleDataUtil`
- [ ] sink
  - [ ] `TestCompressinSettings`
  - [ ] `TestCompressionSink`
  - [ ] `TestFlinkManifest`
  - [ ] `TestIcebergStreamWriter`
  - [ ] `TestIcebergWriter`
  - [ ] `TestRowDataPartitionKey`
  - [ ] `TestTaskWriters`
  - [ ] shuffle
    - [ ] `TestAggregatedStatisticsTracker`

File: flink/v1.19/flink/src/test/java/org/apache/iceberg/flink/source/TestIcebergSourceFailover.java [hunk 1]
```diff
@@ -110,6 +113,14 @@ protected Schema schema() {
     return TestFixtures.SCHEMA;
   }
 
+  protected HadoopTableExtension getSourceTableExtension() {
```
Comment (nastra) [inline flink/v1.19/flink/src/test/java/org/apache/iceberg/flink/source/TestIcebergSourceFailover.java:116]: why are these needed? we typically don't add such methods to tests. Also if this needs to be accessed, then the constant can be accessed directly as it has the same visibility as this method

Comment (tomtongue) [reply flink/v1.19/flink/src/test/java/org/apache/iceberg/flink/source/TestIcebergSourceFailover.java:116]: This is for the inherited class `TestIcebergSourceFailoverWithWatermarkExtractor`. Specifically this is for the tests in this class to use its own `SOURCE_TABLE_EXTENSION` and `SINK_TABLE_EXTENSION` constants. by default, the two constants in the original class `TestIcebergSourceFailover` are used in `TestIcebergSourceFailoverWithWatermarkExtractor` tests.

But I feel the same like your comment. Let me change this part to non-use of these methods.

Comment (tomtongue) [reply flink/v1.19/flink/src/test/java/org/apache/iceberg/flink/source/TestIcebergSourceFailover.java:116]: Pushed the new classes, but still working on these classes.

**Comment by tomtongue** [comment]
@nastra Added all the remaining classes and removed `getSourceTableExtension` and `getSinkTableExtension`. Could you review the changes?

File: flink/v1.19/flink/src/test/java/org/apache/iceberg/flink/sink/TestCompressionSettings.java [hunk 2]
```diff
@@ -91,19 +94,36 @@ public void testCompressionAvro() throws Exception {
...
-          TableProperties.AVRO_COMPRESSION_DEFAULT,
-          resultProperties.get(TableProperties.AVRO_COMPRESSION));
-      Assert.assertEquals(
-          TableProperties.AVRO_COMPRESSION_LEVEL_DEFAULT,
-          resultProperties.get(TableProperties.AVRO_COMPRESSION_LEVEL));
+      assertThat(resultProperties)
+          .containsEntry(TableProperties.AVRO_COMPRESSION, TableProperties.AVRO_COMPRESSION_DEFAULT)
+          .satisfies(
+              props -> {
+                if (props.containsKey(TableProperties.AVRO_COMPRESSION_LEVEL_DEFAULT)) {
```
Comment (nastra) [inline flink/v1.19/flink/src/test/java/org/apache/iceberg/flink/sink/TestCompressionSettings.java:101]: I don't fully follow this assertion here. Isn't this checking whether the map contains a null key because `TableProperties.AVRO_COMPRESSION_LEVEL_DEFAULT` defaults to null?

File: flink/v1.19/flink/src/test/java/org/apache/iceberg/flink/sink/TestCompressionSettings.java [hunk 3]
```diff
@@ -91,19 +94,36 @@ public void testCompressionAvro() throws Exception {
...
-      Assert.assertEquals(
-          initProperties.get(TableProperties.AVRO_COMPRESSION_LEVEL),
-          resultProperties.get(TableProperties.AVRO_COMPRESSION_LEVEL));
+      assertThat(resultProperties)
+          .containsEntry(
+              TableProperties.AVRO_COMPRESSION,
+              initProperties.get(TableProperties.AVRO_COMPRESSION))
+          .satisfies(
+              props -> {
+                if (props.containsKey(TableProperties.AVRO_COMPRESSION_LEVEL)) {
```
Comment (nastra) [inline flink/v1.19/flink/src/test/java/org/apache/iceberg/flink/sink/TestCompressionSettings.java:118]: why is the if/else needed here and not in the original assertion?

File: flink/v1.19/flink/src/test/java/org/apache/iceberg/flink/sink/TestCompressionSettings.java (hunk 2 above)
Comment (tomtongue) [reply flink/v1.19/flink/src/test/java/org/apache/iceberg/flink/sink/TestCompressionSettings.java:101]: Sorry, I'll fix this part. As you're mentioning, this part checks if the compression is null or not null based on a specified compression format.

File: flink/v1.19/flink/src/test/java/org/apache/iceberg/flink/sink/TestCompressionSettings.java (hunk 3 above)
Comment (tomtongue) [reply flink/v1.19/flink/src/test/java/org/apache/iceberg/flink/sink/TestCompressionSettings.java:118]: Sorry this is actually not needed. Will push the fix.

**Review by nastra — APPROVED** [review]

**Comment by tomtongue** [comment]
Thanks for the review! Will submit a backport PR, then the Flink migration will be complete.
